cpp/src/arrow/type.cc

```diff
@@ -501,20 +501,35 @@ Status Decimal128Type::Make(int32_t precision, int32_t scale,
 // ----------------------------------------------------------------------
 // Dictionary-encoded type
 
+Status DictionaryType::ValidateParameters(const DataType& index_type,
+                                          const DataType& value_type) {
+  const bool index_type_ok = is_integer(index_type.id()) &&
+                             checked_cast<const IntegerType&>(index_type).is_signed();
+  if (!index_type_ok) {
+    return Status::TypeError("Dictionary index type should be signed integer, got ",
```

Comment:
Was there any reason that only signed integer is supported and not unsigned ?

Reply:
I'm not sure about the history (@wesm could tell more), but it's specified here:
https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout

> When a field is dictionary encoded, the values are represented by an array of signed integers representing the index of the value in the dictionary.

Reply:
Thank you for documentation.

Reply:
Easier cross language support I believe

Reply:
Yes that was the thinking.

Reply:
cc @trxcllnt who claimed that other Arrow implementations (I presume JS) didn't enforce signed types for the indices

Reply:
This appears to be only mentioned in [Columnar.rst](https://github.com/apache/arrow/blob/master/docs/source/format/Columnar.rst#dictionary-encoded-layout) but not in [Schema.fbs](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L282).  Schema.fbs only mentioned the default is signed.

Reply:
I'm against supporting unsigned dictionary indices. The costs outweigh the benefits in my opinion. If you want to discuss further could we move this to the Arrow dev mailing list?